Add istio external auth tutorial (#13665)
kyma-project/kyma: docs/03-tutorials/00-api-exposure/apix-04-expose-and-secure-workload-istio.md

---
title: Expose and secure a workload with Istio
---

This tutorial shows how to expose a workload with Virtual Service and secure it using [Authorization Policy](https://istio.io/latest/docs/reference/config/security/authorization-policy/) together with [Request Authentication](https://istio.io/latest/docs/reference/config/security/request_authentication/) based on JWT token. 

> 

Reviewer: ```suggestion
This tutorial shows how to expose and secure a workload using Istio build in security features.  You will expose the workload via creating a [Virtual Service](https://istio.io/latest/docs/reference/config/networking/virtual-service/) and securing the access to it via adding JWT token validation verified by Istio security configuration with [Authorization Policy](https://istio.io/latest/docs/reference/config/security/authorization-policy/) together with [Request Authentication](https://istio.io/latest/docs/reference/config/security/request_authentication/) 
```

## Prerequisites

To follow this tutorial, use Kyma 2.0 or higher.

This tutorial is based on a sample HttpBin service deployment and a sample Function. To deploy or create one of those, follow the [Create a workload](./apix-02-create-workload.md) tutorial. It can also be a follow up to the [Use a custom domain to expose a workload](./apix-01-own-domain.md) tutorial.

> ## Get a JWT access token

Reviewer 2: JWT = JSON Web Token, so it already contains the "token" word, doesn't it? 🤔

1. In your OpenID Connect-compliant (OIDC-compliant) identity provider, create an application to get your client credentials such as Client ID and Client Secret. Export your client credentials as environment variables. Run:

   ```bash
   export CLIENT_ID={YOUR_CLIENT_ID}
   export CLIENT_SECRET={YOUR_CLIENT_SECRET}
   ```

2. Encode your client credentials and export them as an environment variable:

   ```bash
   export ENCODED_CREDENTIALS=$(echo -n "$CLIENT_ID:$CLIENT_SECRET" | base64)
   ```

3. In your browser, go to `https://YOUR_OIDC_COMPLIANT_IDENTITY_PROVIDER_INSTANCE/.well-known/openid-configuration`, save values of the **token_endpoint**, **jwks_uri** and **issuer** parameters, and export them as environment variables:

   ```bash
   export TOKEN_ENDPOINT={YOUR_TOKEN_ENDPOINT}
   export JWKS_URI={YOUR_JWKS_URI}
   export ISSUER={YOUR_ISSUER}
   ```

> 4. Get the JWT access token:

Reviewer 2: ```suggestion
4. Get the JWT:
```

Reviewer 2: Same question as above 🤔

   ```bash
   curl -X POST "$TOKEN_ENDPOINT" -d "grant_type=client_credentials" -d "client_id=$CLIENT_ID" -H "Content-Type: application/x-www-form-urlencoded" -H "Authorization: Basic $ENCODED_CREDENTIALS"
   ```

> 5. Save the result, and export it as an environment variable:

Reviewer 2: ```suggestion
5. Save the token and export it as an environment variable:
```

   ```bash
   export ACCESS_TOKEN={YOUR_ACCESSS_TOKEN}
   ```

> ## Expose your workload using VirtualService

Reviewer 2: ```suggestion
## Expose your workload using a Virtual Service
```

> Follow the instructions in the tabs to expose httpbin workload or a function using VirtualService.

Reviewer 2: ```suggestion
Follow the instructions in the tabs to expose the Httpbin workload or a Function using a Virtual Service.
```

<div tabs>

  <details>
  <summary>

>   Expose HttpBin

Reviewer 2: ```suggestion
  Expose the Httpbin workload
```

Reviewer 2: ... or if you prefer HttpBin, correct all the `Httpbin` occurrences so that you are consistent within the document

  </summary>

> 1. Export the following environment variables:

Reviewer 2: ```suggestion
1. Export your domain name and gateway as environment variables:
```

Reviewer 2: ^ or sth like that; anyway, be more precise

   ```shell

>    export DOMAIN_TO_EXPOSE_WORKLOADS={DOMAIN_NAME} # This is a Kyma domain or your custom subdomain e.g. api.mydomain.com.

Reviewer 2: ```suggestion
   export DOMAIN_TO_EXPOSE_WORKLOADS={DOMAIN_NAME} # This is a Kyma domain or your custom subdomain, for example: api.mydomain.com.
```

   export GATEWAY=$NAMESPACE/httpbin-gateway # If you don't want to use your custom domain but a Kyma domain, use the following Kyma Gateway: `kyma-system/kyma-gateway`.
   ```

2. Run:

   ```shell
   cat <<EOF | kubectl apply -f -
   apiVersion: networking.istio.io/v1alpha3
   kind: VirtualService
   metadata:
     name: httpbin
     namespace: $NAMESPACE
   spec:
     hosts:
     - "httpbin.$DOMAIN_TO_EXPOSE_WORKLOADS"
     gateways:
     - $GATEWAY
     http:
     - match:
       - uri:
           prefix: /
       route:
       - destination:
           port:
             number: 8000
           host: httpbin.$NAMESPACE.svc.cluster.local
   EOF
   ```
  </details>

  <details>
  <summary>

>   Expose a function

Reviewer 2: ```suggestion
  Expose a Function
```

  </summary>

> 1. Export the following environment variables:

Reviewer 2: ```suggestion
1. Export your domain name and gateway as environment variables:
```

   ```shell

>    export DOMAIN_TO_EXPOSE_WORKLOADS={DOMAIN_NAME} # This is a Kyma domain or your custom subdomain e.g. api.mydomain.com.

Reviewer 2: ```suggestion
   export DOMAIN_TO_EXPOSE_WORKLOADS={DOMAIN_NAME} # This is a Kyma domain or your custom subdomain, for example: api.mydomain.com.
```

   export GATEWAY=$NAMESPACE/httpbin-gateway # If you don't want to use your custom domain but a Kyma domain, use the following Kyma Gateway: `kyma-system/kyma-gateway`.
   ```

2. Run:

   ```shell
   cat <<EOF | kubectl apply -f -
   apiVersion: networking.istio.io/v1alpha3
   kind: VirtualService
   metadata:
     name: function
     namespace: $NAMESPACE
   spec:
     hosts:
     - "function.$DOMAIN_TO_EXPOSE_WORKLOADS"
     gateways:
     - $GATEWAY
     http:
     - match:
       - uri:
           prefix: /
       route:
       - destination:
           port:
             number: 80
           host: function.$NAMESPACE.svc.cluster.local
   EOF
   ```

  </details>
</div>

> ## Add a RequestAuthentication which requires JWT token for all requests for workloads that have matching label

Reviewer 2: ```suggestion
## Secure a workload or a Function using JWT
```

> Follow the instructions in the tabs to secure httpbin or a function using JWT token.

Reviewer 2: ```suggestion
To secure the Httpbin workload or a Function using JWT, add a Request Authentication. It requires a JWT for all requests for workloads that have matching label. Follow the instructions in the tabs:
```

Reviewer 2: ^ this sentence still needs some polishing, let's discuss

<div tabs>

  <details>
  <summary>

>   Secure HttpBin

Reviewer 2: ```suggestion
  Secure the Httpbin workload
```

  </summary>

> 1. Run:

Reviewer 2: What is the outcome of running this?

> To secure the Httpbin workload or the Function using a JWT, create a Request Authentication with AuthorizationPolicy. Workloads that have the **matchLabels** parameter specified require a JWT for all requests. Follow the instructions in the tabs:

Reviewer 2: ```suggestion
To secure the Httpbin workload or the Function using a JWT, create a Request Authentication with Authorization Policy. Workloads that have the **matchLabels** parameter specified require a JWT for all requests. Follow the instructions in the tabs:
```

   ```shell
   cat <<EOF | kubectl apply -f -
   apiVersion: security.istio.io/v1beta1
   kind: RequestAuthentication
   metadata:
     name: jwt-auth-httpbin
     namespace: $NAMESPACE
   spec:
     selector:
       matchLabels:
         app: httpbin
     jwtRules:
     - issuer: $ISSUER
       jwksUri: $JWKS_URI
   ---
   apiVersion: security.istio.io/v1beta1
   kind: AuthorizationPolicy
   metadata:
     name: httpbin
     namespace: $NAMESPACE
   spec:
     selector:
       matchLabels:
         app: httpbin
     rules:
     - from:
       - source:
           requestPrincipals: ["*"]
   EOF
   ```

> 2. If you try to access secured workload you should get 403 Forbidden error:

Reviewer 2: ```suggestion
2. Access the workload you secured. You will get the `403 Forbidden` error.
```

   ```shell
   curl -ik -X GET https://httpbin.$DOMAIN_TO_EXPOSE_WORKLOADS/status/200
   ```

> 3. Using correct JWT token should give you 200 OK response

Reviewer 2: ```suggestion
3. Now, access the secured workload using the correct JWT. You will get the `200 OK` response.
```

   ```shell
   curl -ik -X GET https://httpbin.$DOMAIN_TO_EXPOSE_WORKLOADS/status/200 --header "Authorization:Bearer $ACCESS_TOKEN"
   ```
  </details>

  <details>
  <summary>

>   Secure a function

Reviewer 2: ```suggestion
  Secure a Function
```

  </summary>

> 1. Run:

Reviewer 2: What is the outcome of running this?

   ```shell
   cat <<EOF | kubectl apply -f -
   apiVersion: security.istio.io/v1beta1
   kind: RequestAuthentication
   metadata:
     name: jwt-auth-function
     namespace: $NAMESPACE
   spec:
     selector:
       matchLabels:
         app: function
     jwtRules:
     - issuer: $ISSUER
       jwksUri: $JWKS_URI
   ---
   apiVersion: security.istio.io/v1beta1
   kind: AuthorizationPolicy
   metadata:
     name: function
     namespace: $NAMESPACE
   spec:
     selector:
       matchLabels:
         app: function
     rules:
     - from:
       - source:
           requestPrincipals: ["*"]
   EOF
   ```

> 2. If you try to access secured workload you should get 403 Forbidden error:

Reviewer 2: ```suggestion
2. Access the workload you secured. You will get the `403 Forbidden` error.
```

   ```shell
   curl -ik -X GET https://function.$DOMAIN_TO_EXPOSE_WORKLOADS/status/200
   ```

> 3. Using correct JWT token should give you 200 OK response

Reviewer 2: ```suggestion
3. Now, access the secured workload using the correct JWT. You will get the `200 OK` response.
```

   export CLIENT_SECRET={YOUR_CLIENT_SECRET}
   ```

2. Encode your client credentials and export them as an environment variable:

   ```bash
   export ENCODED_CREDENTIALS=$(echo -n "$CLIENT_ID:$CLIENT_SECRET" | base64)
   ```

3. In your browser, go to `https://YOUR_OIDC_COMPLIANT_IDENTITY_PROVIDER_INSTANCE/.well-known/openid-configuration`, save values of the **token_endpoint** and **jwks_uri** parameters, and export them as environment variables:

   ```bash
   export TOKEN_ENDPOINT={YOUR_TOKEN_ENDPOINT}
   export JWKS_URI={YOUR_JWKS_URI}
   ```

4. Get the JWT access token:

   ```bash
   curl -X POST "$TOKEN_ENDPOINT" -d "grant_type=client_credentials" -d "client_id=$CLIENT_ID" -H "Content-Type: application/x-www-form-urlencoded" -H "Authorization: Basic $ENCODED_CREDENTIALS"
   ```

5. Save the result, and export it as an environment variable:

   ```bash
   export ACCESS_TOKEN={YOUR_ACCESSS_TOKEN}
   ```

## Expose your workload using VirtualService

Follow the instructions in the tabs to expose httpbin workload or a function using VirtualService.

<div tabs>

  <details>
  <summary>
  Expose HttpBin
  </summary>

1. Export the following environment variables:

   ```shell

>    export DOMAIN={DOMAIN_NAME} # This is a Kyma domain or your custom subdomain e.g. api.mydomain.com.

Reviewer 3: in other tutorials we are using 
```suggestion
   export DOMAIN_TO_EXPOSE_WORKLOADS={DOMAIN_NAME} #This is a Kyma domain or your custom subdomain e.g. api.mydomain.com.
```

Reviewer 3: please adjust it everywhere in this tut

   spec:
     hosts:
     - "httpbin.$DOMAIN"
     gateways:
     - $GATEWAY
     http:
     - match:
       - uri:
           prefix: /
       route:
       - destination:
           port:
             number: 8000
           host: httpbin.$NAMESPACE.svc.cluster.local
   EOF
   ```
  </details>

  <details>
  <summary>
  Expose a function
  </summary>

1. Export the following environment variables:

   ```shell
   export DOMAIN={DOMAIN_NAME} # This is a Kyma domain or your custom subdomain e.g. api.mydomain.com.
   export GATEWAY=$NAMESPACE/httpbin-gateway # If you don't want to use your custom domain but a Kyma domain, use the following Kyma Gateway: `kyma-system/kyma-gateway`.
   ```

1. Run:

   ```shell
   cat <<EOF | kubectl apply -f -
   apiVersion: networking.istio.io/v1alpha3
   kind: VirtualService
   metadata:
     name: function
     namespace: $NAMESPACE
   spec:
     hosts:
     - "function.$DOMAIN"
     gateways:
     - $GATEWAY
     http:
     - match:
       - uri:
           prefix: /
       route:
       - destination:
           port:
             number: 80
           host: function.$NAMESPACE.svc.cluster.local
   EOF
   ```

  </details>
</div>

> ## Add a RequestAuthentication which requires JWT token for all requests for workloads that have label app:httpbin

Reviewer 3: ```suggestion
## Add a RequestAuthentication which requires JWT token for all requests for workloads that have matching label 
```

Follow the instructions in the tabs to secure httpbin or a function using JWT token.

<div tabs>

  <details>
  <summary>
  Secure HttpBin
  </summary>

1. Export the following values:

   ```shell
   export ISSUER={ISSUER} # e.g. https://YOUR_OIDC_COMPLIANT_IDENTITY_PROVIDER_INSTANCE

>    export JWKS_URI={JWKS_URL} # e.g. https://YOUR_OIDC_COMPLIANT_IDENTITY_PROVIDER_INSTANCE/.well-known/jwks.json

Reviewer 3: this is already exported. maybe it makes sense to move issuer export too?

   ```

1. Run:

   ```shell
   cat <<EOF | kubectl apply -f -
   apiVersion: security.istio.io/v1beta1
   kind: RequestAuthentication
   metadata:
     name: jwt-auth-httpbin
     namespace: $NAMESPACE
   spec:
     selector:
       matchLabels:
         app: httpbin
     jwtRules:
     - issuer: $ISSUER
       jwksUri: $JWKS_URI
   ---
   apiVersion: security.istio.io/v1beta1
   kind: AuthorizationPolicy
   metadata:
     name: httpbin
     namespace: $NAMESPACE
   spec:
     selector:
       matchLabels:
         app: httpbin
     rules:
     - from:
       - source:
           requestPrincipals: ["*"]
   EOF
   ```

2. If you try to access secured workload you should get 403 Forbidden error:

   ```shell
   curl -ik -X GET https://httpbin.$DOMAIN/status/200
   ```

3. Using correct JWT token should give you 200 OK response

   ```shell
   curl -ik -X GET https://httpbin.$DOMAIN/status/200 --header "Authorization:Bearer $ACCESS_TOKEN"
   ```
  </details>

  <details>
  <summary>
  Secure a function
  </summary>

1. Export the following values:

   ```shell
   export ISSUER={ISSUER} # e.g. https://YOUR_OIDC_COMPLIANT_IDENTITY_PROVIDER_INSTANCE.com

>    export JWKS_URI={JWKS_URL} # e.g. https://YOUR_OIDC_COMPLIANT_IDENTITY_PROVIDER_INSTANCE/.well-known/jwks.json

Reviewer 3: this is already exported. maybe it makes sense to move issuer export too?
